### PR TITLE
TypeScript Type Declarations

### PR DESCRIPTION
**Pull request by jdconley**
This includes type declarations for use automatically by TypeScript 2.0, some minor edits to the README to reflect how to use the library in TypeScript, and an update to the package.json which tells the TypeScript compiler where to find typings for the module.

This is based on the most recent documentation, here: 
https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-d-ts.html

and here: 
https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html

I could also just add these to the DefinitelyTyped project, but thought it would be nice to be local with the library since that is supported now out of the box for TypeScript and it's just a tiny extra file to include in this module.

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+export interface IOptionLimits {
+    hashLength: INumericLimit;
+    memoryCost: INumericLimit;
+    timeCost: INumericLimit;
+    parallelism: INumericLimit;
+}
+
+export const defaults: IOptions;
+export const limits: IOptionLimits;
+export function hash(plain: string, salt: Buffer | string, options?: IOptions): Promise<string>;
```
Comment (ranisalt) [inline index.d.ts]: Salt must be a Buffer, password (plain) can be either a buffer or a string (it's converted anyway)

**Comment by ranisalt** [comment]
Why changing single quotes to double?

**Comment by jdconley** [comment]
The quotes are a TypeScript style guide thing. The linter fails by default with single quoted strings. I can change the quotes back if you want. https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#strings

Though I just realized I broke another TypeScript style by using the "I" prefix on the interfaces. I'm going to change those in the declarations.

**Comment by ranisalt** [comment]
It's because you only changed them on the examples, doesn't it matter that index.js uses single quotes?

**Comment by jdconley** [comment]
The double quotes are just for when you are linting the code with the "tslint" tool and subsequently transpiling with the TypeScript transpiler. I just thought the examples should match what the TypeScript tooling expects by default. Single quotes are legal. The linter settings just need to be adjusted.

In this case we are just supplying the types for your library to TypeScript, not compiling it with TypeScript, so the implementation in your JS file doesn't matter.

File: index.d.ts (hunk 1 above)
Comment (jdconley) [inline index.d.ts]: I just updated the data types in commit 3895ab9.

Comment (jdconley) [inline index.d.ts]: And 4ff805a for salt param. Did that too fast and didn't double check before I committed...

**Comment by jdconley** [comment]
@ranisalt do you want single quotes in the README? It looks like the build failed intermittently because `npm install` timed out on Travis CI, not because of these changes...

**Comment by ranisalt** [comment]
Travis recently is driving me nuts, it fails unpredictably on random projects.

**Comment by jdconley** [comment]
I updated the README with a TypeScript section and removed instructions from the type declaration file -- since 2.0 is still pretty new I thought some extra assistance would be useful.

I fixed a bug in the typings (length is optional for generateSalt).

I added a reference to the "node" typings and a devDependency as well, which is where the Buffer interface comes from in this type declaration file. I didn't want to assume that consumers have node typings loaded globally in the compiler.

In addition, I added a tsconfig.json which instructs the typescript compiler to assume Promises are available globally. This is primarily so that someone authoring the index.d.ts file in a smart IDE like VS Code or IDEA will not see warnings about "Promise" being an unknown type.

File: README.md [hunk 2]
```diff
@@ -139,6 +139,38 @@ not raw.
...
+
+{
+    "compilerOptions": {
+        "target": "es6"
+    }
+}
+```
+
+```ts
+import * as argon2 from "argon2";
```
Comment (ranisalt) [inline README.md]: Question, in ES2015 you would just `import argon2 from 'argon2'`, is this wildcard really necessary with TS?

Comment (jdconley) [inline README.md]: You could `import { argon2 } from "argon2";` in TypeScript if you only wanted access to that interface, yes. The import * just brings in all exported things from that module. In order to use the `Options` interface on the hash method, for instance, you would need to import it as well.

**Comment by ranisalt** [comment]
Could you make tests to check if the typing interface is properly set up?

**Comment by jdconley** [comment]
I just started using these type declarations in another project that isn't yet public, and they are working. I was just researching the best way to automatically validate the typing interface. Unfortunately, there isn't really a great way to do that. One way would be to run your tests through TypeScript. From a cursory look it seems like they are very close to working out of the box.

If you like TypeScript I could convert the whole module to TypeScript, including the tests, very easily since most of your code is in the C++. If your code is TypeScript there is a function built-in to the transpiler that will generate the declaration file so it is guaranteed to be accurate.

**Comment by jdconley** [comment]
I just added a test to `npm test` in 9fabaa3 that verifies the interface. It just runs the TypeScript compiler (tsc), not the code. It's the same way they do it in the [DefinitelyTyped](http://definitelytyped.org/guides/contributing.html) project. I created a truncated version of your tests that test all the various interface options.

**Pull request closed without merging**